metrics/metric_reporter.go

```
@@ -0,0 +1,139 @@
+type fetchedMetric struct {
+	key   string
+	value string
+	tags  string
+}
+
+func NewMetricReporter() *MetricReporter {
+	m := MetricReporter{}
+	reg := prometheus.NewRegistry()
+	m.reporter = promreporter.NewReporter(promreporter.Options{Registerer: reg})
+
+	m.scopes = make(map[string]tally.Scope)
+	m.scopes["root"], m.closer = tally.NewRootScope(tally.ScopeOptions{
+		Tags:           GetBaseTags(),
+		CachedReporter: m.reporter,
+		Separator:      promreporter.DefaultSeparator,
+	}, 1*time.Second)
+
+	m.AddScope(m.scopes["root"], "fdb", "fdb")
+	m.AddScope(m.scopes["fdb"], "client", "client")
+	m.AddScope(m.scopes["fdb"], "cluster", "cluster")
+	m.AddScope(m.scopes["cluster"], "workload", "workload")
+
+	// Add each implemented group here
+	m.groups = []Collectable{
+		NewCoordinatorMetricGroup(&m),
+		NewDbStatusMetricGroup(&m),
+		NewWorkloadOperationsMetricGroup(&m),
+		NewWorkloadTransactionsMetricGroup(&m),
+		NewWorkloadKeysMetricGroup(&m),
+		NewWorkloadBytesMetricGroup(&m),
+		NewDataMetricGroup(&m),
+		NewProcessesMetricGroup(&m),
+		NewLatencyProbeMetricGroup(&m),
+		NewBackupMetricGroup(&m),
+		NewClusterMessageMetricGroup(&m),
+	}
+	return &m
+}
+
+// Single data collection, fetches status, gets the metrics from each group
+func (m *MetricReporter) collectOnce() error {
+	var err error
+	m.status, err = db.GetStatus()
+	if err != nil {
+		return fmt.Errorf("failed to get status")
+	}
+
+	if len(m.groups) == 0 {
+		ulog.E(fmt.Errorf("no metric groups detected"))
+	}
+
+	for _, group := range m.groups {
+		group.GetMetrics(m.status)
+	}
+	return nil
+}
+
+// Used only in integration testing, collects metrics from a json file
+func (m *MetricReporter) collectOnceFromFile(fileName string) error {
+	// Used in testing
+	wd, err := os.Getwd()
+	if err != nil {
+		ulog.E(err)
+	}
+	testFilePath := fmt.Sprintf("%s/../%s/%s", wd, RelativeJsonFileLocation, fileName)
+	f, err := os.Open(testFilePath)
+	if err != nil {
+		ulog.E(err)
```

If not able to open, then we need to return an error.